src/modules/powerrename/ui/PowerRenameUI.cpp

```diff
@@ -0,0 +1,930 @@
+
+    return S_OK;
+}
+
+IFACEMETHODIMP CPowerRenameUI::DragOver(DWORD /* grfKeyState */, POINTL pt, _Inout_ DWORD* pdwEffect)
+{
+    if (m_spdth)
+    {
+        POINT ptT = { pt.x, pt.y };
+        m_spdth->DragOver(&ptT, *pdwEffect);
+    }
+
+    return S_OK;
+}
+
+IFACEMETHODIMP CPowerRenameUI::DragLeave()
+{
+    if (m_spdth)
+    {
+        m_spdth->DragLeave();
+    }
+
+    return S_OK;
+}
+
+IFACEMETHODIMP CPowerRenameUI::Drop(_In_ IDataObject* pdtobj, DWORD, POINTL pt, _Inout_ DWORD* pdwEffect)
+{
+    if (m_spdth)
+    {
+        POINT ptT = { pt.x, pt.y };
+        m_spdth->Drop(pdtobj, &ptT, *pdwEffect);
+    }
+
+    _OnClear();
+
+    EnableWindow(GetDlgItem(m_hwnd, ID_RENAME), TRUE);
+    EnableWindow(m_hwndLV, TRUE);
+
+    // Populate the manager from the data object
+    if (m_spsrm)
+    {
+        _EnumerateItems(pdtobj);
+    }
+
+    return S_OK;
+}
+
+HRESULT CPowerRenameUI::_Initialize(_In_ IPowerRenameManager* psrm, _In_opt_ IDataObject* pdo, _In_ bool enableDragDrop)
+{
+    // Cache the smart rename manager
+    m_spsrm = psrm;
+
+    // Cache the data object for enumeration later
+    m_spdo = pdo;
+
+    m_enableDragDrop = enableDragDrop;
+
+    HRESULT hr = CoCreateInstance(CLSID_DragDropHelper, NULL, CLSCTX_INPROC, IID_PPV_ARGS(&m_spdth));
+    if (SUCCEEDED(hr))
+    {
+        // Subscribe to smart rename manager events
+        hr = m_spsrm->Advise(this, &m_cookie);
+    }
+
+    if (FAILED(hr))
+    {
+        _Cleanup();
+    }
+
+    return hr;
+}
+
+void CPowerRenameUI::_Cleanup()
+{
+    if (m_spsrm && m_cookie != 0)
+    {
+        m_spsrm->UnAdvise(m_cookie);
+        m_cookie = 0;
+        m_spsrm = nullptr;
+    }
+
+    m_spdo = nullptr;
+    m_spdth = nullptr;
+
+    if (m_enableDragDrop)
+    {
+        RevokeDragDrop(m_hwnd);
+    }
+}
+
+void CPowerRenameUI::_EnumerateItems(_In_ IDataObject* pdtobj)
+{
+    // Enumerate the data object and popuplate the manager
+    if (m_spsrm)
+    {
+        m_disableCountUpdate = true;
+        EnumerateDataObject(pdtobj, m_spsrm);
+        m_disableCountUpdate = false;
+
+        UINT itemCount = 0;
+        m_spsrm->GetItemCount(&itemCount);
+        m_listview.SetItemCount(itemCount);
+
+        _UpdateCounts();
+    }
+}
+
+// TODO: persist settings made in the UI
+HRESULT CPowerRenameUI::_ReadSettings()
+{
+    return S_OK;
+}
+
+HRESULT CPowerRenameUI::_WriteSettings()
+{
+    return S_OK;
+}
+
+void CPowerRenameUI::_OnClear()
+{
+}
+
+void CPowerRenameUI::_OnCloseDlg()
+{
+    // Persist the current settings
+    _WriteSettings();
+    EndDialog(m_hwnd, 1);
+}
+
+void CPowerRenameUI::_OnDestroyDlg()
+{
+    _Cleanup();
+}
+
+void CPowerRenameUI::_OnRename()
+{
+    if (m_spsrm)
+    {
+        m_spsrm->Rename(m_hwnd);
+    }
+}
+
+void CPowerRenameUI::_OnAbout()
+{
+    // Launch github page
+    SHELLEXECUTEINFO info = {0};
+    info.cbSize = sizeof(SHELLEXECUTEINFO);
+    info.lpVerb = L"open";
+    info.lpFile = L"http://www.github.com/chrdavis";
```

Comment:
Maybe this should point to https://github.com/microsoft/PowerToys/tree/master/src/modules/powerrename?